cilium: Small health cleanup improvements

borkmann opened the pull request:
(see commit msg)

Fixes: https://github.com/cilium/cilium/issues/32689

rgo3 committed:
cilium: Remove health ep from ep manager on shutdown

Currently on shutdown we don't remove the health endpoint from the
endpoint manager but do remove the corresponding netlink interface.
Note that validateEndpoint() special cases the health endpoint and
does remove the old state directory if it existed before. But
independent of that it would be better to just streamline cleanup
via cleanupHealthEndpoint() for all cases.

Related: #32689
Signed-off-by: Robin Gögge <r.goegge@isovalent.com>
Co-developed-by: Daniel Borkmann <daniel@iogearbox.net>
Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

rgo3 committed:
gh: e2e-upgrade: enable setup 7

Reenable setup 7 CI run in upgrade test.

Signed-off-by: Robin Gögge <r.goegge@isovalent.com>
Co-developed-by: Daniel Borkmann <daniel@iogearbox.net>
Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium: Add a log for cleaning up health endpoint

We log whenever we start the health endpoint. Also add a log entry
whenever we attempt to clean up.

Also add a small comment in the cleanup routine for device removal.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

gandro commented on daemon/cmd/state.go:
```diff
@@ -82,6 +83,8 @@ func (d *Daemon) validateEndpoint(ep *endpoint.Endpoint) (valid bool, err error)
 		if err := os.RemoveAll(healthStateDir); err != nil {
 			scopedLog.Warning("Cannot clean up old health state directory")
 		}
+		// Remove health related devices if they still exist
```
After re-reading the code, this seems like the proper place to do cleanup for now.

I think we should invoke `health.KillEndpoint()` here too. Otherwise we try to remove the network namespace of a running process, which I'm not sure works well.

gandro reviewed (commented):
I think this looks good. Have you tested if this fixes CI?

gandro commented on daemon/cmd/health.go:
```diff
@@ -147,5 +147,4 @@ func (d *Daemon) cleanupHealthEndpoint() {
 			log.WithError(err).Debug("Error occurred while deleting cilium-health endpoint")
 		}
 	}
-	health.CleanupEndpoint()
```
We should keep this here, otherwise the health device might be leaked if cilium is never started again

gandro reviewed (approved):
Code looks good to me. We should also test if this really fixes CI like the previous PR did: https://github.com/cilium/cilium/pull/33410/commits/dd94c27a4741bf0e56225f9b65d9763783875475

borkmann committed:
cilium: Reuse common endpoint deletion functionality

Do not open code it in the health code.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium: Trigger reping after launching the health endpoint

Otherwise we might run risk of hitting the timeout too early. Just reping
right away upon success.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium: Add a comment on device management

For the health endpoint the agent needs to remove devices since its not
managed by CNI. Add a comment explaining this.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

tommyp1ckles committed:
endpoint: hold build mutex during delete.

The endpoints buildMutex is commented as intending to be
held while doing endpoint build as well as delete.

However, this lock is not actually held during (*Endpoint).Delete().
As a result, there is a race condition that can occur where an endpoint
in the process of doing `e.owner.GetCompilationLock().RLock()` may be
deleted while waiting for this lock.

This is made more likely as the write lock on the compilation lock can
be held for several minutes at times.

When the regenerateBPF finally does acquire the lock, it may find that
it's underlying interface has already been deleted.  Specifically, in
the case of the health endpoints lxc_health interface, this is removed
after some ping timeout and endpoint-manager is called to delete the
endpoint.  However if the endpoint waiting for the compilation lock
at the same time it will still attempt to regenerate the endpoint once
the rlock becomes available, leading to errors such as:

```
msg="Error while reloading endpoint BPF program" ciliumEndpointName=/
containerID= containerInterface= datapathPolicyRevision=0
desiredPolicyRevision=1 endpointID=62 error="retrieving device
lxc_health: Link not found" identity=4 ipv4=10.244.0.18
ipv6="fd00:10:244::9647" k8sPodName=/ subsys=endpoint
```

This adds the build mutex locking to (*Endpoint).Delete(), such that
the endpoint cannot deleted while it is being regenerated.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

borkmann committed:
cilium: Extend endpoint deletion to wait for any ongoing builds to finish

Endpoint deletions can race with pending builds as detailed in #32689.
This is due to the fact that the Delete() call does not wait for them
to finish. Instead, it tears down the endpoint, downs the device, removes
tc(x), etc. This is not correct since the inflight from regenerate() can
later still attempt to attach something to the given device. This in itself
is not problematic as the CNI will just remove the device and the kernel
removes all objects along with it. However, this can result in sporadic
agent error logs that a given lxc* device has already disappeared when
the agent tries to attach, but Delete() resp the CNI Del finished. We've
seen this in particular in case of lxc_health. The health endpoint is not
managed by the CNI plugin, but the agent instead. What appears to happen
is that when health endpoint is cleaned, it's removed from the ep manager,
and subsequently the device is removed.. however any ongoing regenerate()
later tries to attach to a non-existing device. Given the Delete() updates
the endpoint state, we also need to wrap this with the build mutex lock.
This is to wait for any pending regenerate() to finish before setting the
new ep state, and any new build requests will bail out early as they see
that the endpoint is disconnecting. Kudos to Tom Hadlaw and Robin Goegge!

Fixes: #32689
Co-developed-by: Tom Hadlaw <tom.hadlaw@isovalent.com>
Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>
Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

ldelossa reviewed (approved):
LGTM

joestringer commented on pkg/endpoint/endpoint.go:
```diff
@@ -2474,6 +2474,12 @@ func (e *Endpoint) Delete(conf DeleteConfig) []error {
+	// Wait for any pending endpoint regenerate() calls to finish. The
+	// latter bails out after taking the lock when it detects that the
+	// endpoint state is disconnecting.
+	e.buildMutex.Lock()
+	defer e.buildMutex.Unlock()
```
Given the comment above `leaveLocked()` I'm surprised this was missing from this path. The lock ordering here looks correct compared to other usage of `buildMutex` (vs. `e.mutex`). It's hard to fully understand the impact of this with deletions potentially blocking more often depending on other endpoint creation/regeneration events going on in the system, but it does seem to be more correct behaviour.
